src/pipeline/src/etl/transform/transformer/greptime.rs

```diff
@@ -420,22 +461,84 @@ fn identity_pipeline_inner<'a>(
 pub fn identity_pipeline(
     array: Vec<serde_json::Value>,
     table: Option<Arc<table::Table>>,
+    params: GreptimeIdentityPipelineParams,
 ) -> Result<Rows> {
     match table {
         Some(table) => {
             let table_info = table.table_info();
             let tag_column_names = table_info.meta.row_key_column_names();
-            identity_pipeline_inner(array, Some(tag_column_names))
+            identity_pipeline_inner(array, Some(tag_column_names), params)
+        }
+        None => identity_pipeline_inner(array, None::<std::iter::Empty<&String>>, params),
+    }
+}
+
+/// Consumes the JSON object and consumes it into a single-level object.
+///
+/// The `max_nested_levels` parameter is used to limit the nested levels of the JSON object.
+/// The error will be returned if the nested levels is greater than the `max_nested_levels`.
+pub fn flatten_json_object(
+    object: Map<String, JsonValue>,
+    max_nested_levels: usize,
+) -> Result<Map<String, JsonValue>> {
+    let mut flattened = Map::new();
+
+    if !object.is_empty() {
+        // it will use recursion to flatten the object.
+        do_flatten_json_object(&mut flattened, None, object, 1, max_nested_levels)?;
+    }
+
+    Ok(flattened)
+}
+
+fn do_flatten_json_object(
+    dest: &mut Map<String, JsonValue>,
+    base: Option<&str>,
+    object: Map<String, JsonValue>,
+    current_level: usize,
+    max_nested_levels: usize,
+) -> Result<()> {
+    // For safety, we do not allow the depth to be greater than the max_object_depth.
+    if current_level > max_nested_levels {
+        return ReachedMaxNestedLevelsSnafu { max_nested_levels }.fail();
+    }
+
+    for (key, value) in object {
+        let new_key = base.map_or_else(|| key.clone(), |base_key| format!("{base_key}.{key}"));
+
+        match value {
+            JsonValue::Object(object) => {
+                do_flatten_json_object(
+                    dest,
+                    Some(&new_key),
+                    object,
+                    current_level + 1,
+                    max_nested_levels,
+                )?;
+            }
+            // To simplify the process of logs collection scenario, we will convert the array into a string that can be easily handled by full-text search.
+            JsonValue::Array(array) => {
+                let array_str = serde_json::to_string(&array).context(SerializeToJsonSnafu {
```

Comment:
Since we have the JSON datatype, can we store the array as JSON array directly instead of `serde_json::to_string`?

Reply:
+1